### PR TITLE
Start kwok controller with static pod when the runtime is kind

### LINKED ISSUE #98: [kwokctl] Deploy `kwok` with a static pod at `kind` runtime
### What would you like to be added?

Deploy `kwok` with a static pod at `kind` runtime

### Why is this needed?

No longer dependent on `controller-manager` at startup

### PR DESCRIPTION
**Pull request by Fish-pro**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:

Fixes #98 

#### Does this PR introduce a user-facing change?

```release-note
Run kwok controller with static pod when the runtime is kind by kwokctl create cluster
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by Fish-pro** [comment]
Here is the test verification process
```sh
➜  ~ kwokctl create cluster --name kwok --runtime kind --kind-node-image kindest/node:v1.21.1
Creating cluster                                                    cluster=kwok
Starting cluster                                                    cluster=kwok
Creating cluster "kwok-kwok" ...
 ✓ Ensuring node image (kindest/node:v1.21.1) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kwok-kwok"
You can now use your cluster with:

kubectl cluster-info --context kind-kwok-kwok

Have a nice day! 👋
Cluster is ready                                                    cluster=kwok
You can now use your cluster with:

    kubectl config use-context kwok-kwok

Thanks for using kwok!
➜  ~ kubectl config use-context kwok-kwok
Switched to context "kwok-kwok".
➜  ~ k -n kube-system get po | grep kwok-controller
kwok-controller-kwok-kwok-control-plane           1/1     Running   0          33s
➜  ~ >....
      key: kwok.x-k8s.io/node
      value: fake
status:
  allocatable:
    cpu: 32
    memory: 256Gi
    pods: 110
  capacity:
    cpu: 32
    memory: 256Gi
    pods: 110
  nodeInfo:
    architecture: amd64
    bootID: ""
    containerRuntimeVersion: ""
    kernelVersion: ""
    kubeProxyVersion: fake
    kubeletVersion: fake
    machineID: ""
    operatingSystem: linux
    osImage: ""
    systemUUID: ""
  phase: Running
EOF
node/kwok-node-0 created
➜  ~ kubectl get node -o wide
NAME                      STATUS                     ROLES                  AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE       KERNEL-VERSION     CONTAINER-RUNTIME
kwok-kwok-control-plane   Ready,SchedulingDisabled   control-plane,master   91s   v1.21.1   172.19.0.2    <none>        Ubuntu 21.04   5.10.76-linuxkit   containerd://1.5.2
kwok-node-0               Ready                      agent                  9s    fake      172.19.0.2    <none>        <unknown>      <unknown>          <unknown>
➜  ~ >....
  template:
    metadata:
      labels:
        app: fake-pod
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
              - matchExpressions:
                  - key: type
                    operator: In
                    values:
                      - kwok
      # A taints was added to an automatically created Node.
      # You can remove taints of Node or add this tolerations.
      tolerations:
        - key: "kwok.x-k8s.io/node"
          operator: "Exists"
          effect: "NoSchedule"
      containers:
        - name: fake-container
          image: fake-image
EOF
deployment.apps/fake-pod created
➜  ~ k get po
NAME                        READY   STATUS    RESTARTS   AGE
fake-pod-759cdcff4c-8hzjh   1/1     Running   0          4s
fake-pod-759cdcff4c-9r2rz   1/1     Running   0          4s
fake-pod-759cdcff4c-cfbx4   1/1     Running   0          4s
fake-pod-759cdcff4c-ckscn   1/1     Running   0          4s
fake-pod-759cdcff4c-fgmnh   1/1     Running   0          4s
fake-pod-759cdcff4c-l75vx   1/1     Running   0          4s
fake-pod-759cdcff4c-qkp85   1/1     Running   0          4s
fake-pod-759cdcff4c-qts55   1/1     Running   0          4s
fake-pod-759cdcff4c-w5mnr   1/1     Running   0          4s
fake-pod-759cdcff4c-w5wmn   1/1     Running   0          4s
```

File: pkg/kwokctl/runtime/kind/kwok_controller_pod.go [hunk 1]
```diff
@@ -25,26 +25,26 @@ import (
 	_ "embed"
 )
 
-//go:generate kubectl kustomize -o kwok_controller_deployment.yaml.tpl .
+//go:generate kubectl kustomize -o kwok_controller_pod.yaml.tpl .
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/kind/kwok_controller_pod.go:28]:
```suggestion
```

Need remove `kustomization.yaml` from the same directory

File: pkg/kwokctl/runtime/kind/kwok_controller_pod.yaml.tpl [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    - mountPath: /etc/kubernetes/admin.conf
+      name: kubeconfig
+      readOnly: true
+  hostNetwork: true
+  restartPolicy: Always
+  volumes:
+  - hostPath:
+      path: /etc/kubernetes/admin.conf
+      type: FileOrCreate
+    name: kubeconfig
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/kind/kwok_controller_pod.yaml.tpl:56]: Add a new line in ends

Comment (wzshiming) [reply pkg/kwokctl/runtime/kind/kwok_controller_pod.yaml.tpl:56]: I was wondering why the checksum passed, but it turns out that the '.tpl' suffix was not checked

https://github.com/kubernetes-sigs/kwok/blob/c6ae593ff5c39497ad44c0bcef09431ca6bd9421/hack/verify-ends-newline.sh#L22-L30

**Review by wzshiming — CHANGES_REQUESTED** [review]

File: pkg/kwokctl/runtime/kind/cluster.go [hunk 3]
```diff
@@ -199,9 +199,8 @@ func (c *Cluster) Up(ctx context.Context) error {
 	if err != nil {
 		return err
 	}
-	err = c.Kubectl(ctx, utils.IOStreams{
-		ErrOut: os.Stderr,
-	}, "apply", "-f", utils.PathJoin(conf.Workdir, runtime.KwokDeploy))
+
+	err = utils.Exec(ctx, "", utils.IOStreams{}, "docker", "cp", utils.PathJoin(conf.Workdir, runtime.KwokPod), conf.Name+"-control-plane:/etc/kubernetes/manifests/")
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/kind/cluster.go:203]: https://github.com/kubernetes-sigs/kwok/blob/c6ae593ff5c39497ad44c0bcef09431ca6bd9421/pkg/kwokctl/runtime/kind/cluster.go#L276-L298

It looks like Stop and Start cannot operate on the Kwok-controller.

```suggestion
	err = utils.Exec(ctx, "", utils.IOStreams{}, "docker", "cp", utils.PathJoin(conf.Workdir, runtime.KwokPod), conf.Name+"-control-plane:/etc/kubernetes/manifests/kwok-controller.yaml")
```

Or modify `runtime.KwokPod`

**Review by wzshiming — CHANGES_REQUESTED** [review]

File: hack/verify-ends-newline.sh [hunk 4]
```diff
@@ -25,6 +25,7 @@ function check_ends() {
     -o -iname "*.sh" \
     -o -iname "*.go" \
     -o -iname "*.yaml" \
+    -o -iname "*.tpl" \
```
Comment (wzshiming) [inline hack/verify-ends-newline.sh:28]: I think it would be better to solve this problem with a new PR

**Review by wzshiming — CHANGES_REQUESTED** [review]

File: pkg/kwokctl/runtime/kind/kwok_controller_pod.go (hunk 1 above)
Comment (Fish-pro) [reply pkg/kwokctl/runtime/kind/kwok_controller_pod.go]: modified

File: pkg/kwokctl/runtime/kind/kwok_controller_pod.yaml.tpl (hunk 2 above)
Comment (Fish-pro) [reply pkg/kwokctl/runtime/kind/kwok_controller_pod.yaml.tpl]: modified

File: pkg/kwokctl/runtime/kind/cluster.go (hunk 3 above)
Comment (Fish-pro) [reply pkg/kwokctl/runtime/kind/cluster.go]: I agree with you kwok-controller.yaml,modified

Comment (Fish-pro) [reply pkg/kwokctl/runtime/kind/cluster.go]: @wzshiming Start and Stop In this mode, whether to stop when moving to another folder?

File: hack/verify-ends-newline.sh (hunk 4 above)
Comment (Fish-pro) [reply hack/verify-ends-newline.sh:28]: ok

Comment (Fish-pro) [reply hack/verify-ends-newline.sh]: @wzshiming I tested it and when `docker exec -it kwok-kwok-control-plane mv /etc/kubernetes/manifests/kube-controller-manager.yaml /etc/kubernetes/manifests/kube-controller-manager.yaml.bak` is executed, it deletes the old pod and starts a new pod
```sh
➜  ~ k -n kube-system get po
NAME                                              READY   STATUS    RESTARTS   AGE
coredns-558bd4d5db-c4sxf                          0/1     Pending   0          15m
coredns-558bd4d5db-x7wlq                          0/1     Pending   0          15m
etcd-kwok-kwok-control-plane                      1/1     Running   0          15m
kindnet-97jkb                                     1/1     Running   0          13m
kindnet-zrtg6                                     1/1     Running   0          15m
kube-apiserver-kwok-kwok-control-plane            1/1     Running   0          15m
kube-controller-manager-kwok-kwok-control-plane   1/1     Running   0          15m
kube-proxy-2f2mg                                  1/1     Running   0          15m
kube-proxy-hclf7                                  1/1     Running   0          13m
kube-scheduler-kwok-kwok-control-plane            1/1     Running   0          15m
kwok-controller-kwok-kwok-control-plane           1/1     Running   0          15m
➜  ~ docker exec -it kwok-kwok-control-plane mv /etc/kubernetes/manifests/kube-controller-manager.yaml /etc/kubernetes/manifests/kube-controller-manager.yaml.bak
➜  ~ k -n kube-system get po
NAME                                              READY   STATUS    RESTARTS   AGE
coredns-558bd4d5db-c4sxf                          0/1     Pending   0          17m
coredns-558bd4d5db-x7wlq                          0/1     Pending   0          17m
etcd-kwok-kwok-control-plane                      1/1     Running   0          17m
kindnet-97jkb                                     1/1     Running   0          16m
kindnet-zrtg6                                     1/1     Running   0          17m
kube-apiserver-kwok-kwok-control-plane            1/1     Running   0          17m
kube-controller-manager-kwok-kwok-control-plane   1/1     Running   0          29s
kube-proxy-2f2mg                                  1/1     Running   0          17m
kube-proxy-hclf7                                  1/1     Running   0          16m
kube-scheduler-kwok-kwok-control-plane            1/1     Running   0          17m
kwok-controller-kwok-kwok-control-plane           1/1     Running   0          17m
```

File: pkg/kwokctl/runtime/kind/cluster.go (hunk 3 above)
Comment (wzshiming) [reply pkg/kwokctl/runtime/kind/cluster.go]: > Start and Stop In this mode, whether to stop when moving to another folder?

Yes

**Comment by Fish-pro** [comment]
@wzshiming I tested it and when `docker exec -it kwok-kwok-control-plane mv /etc/kubernetes/manifests/kube-controller-manager.yaml /etc/kubernetes/manifests/kube-controller-manager.yaml.bak` is executed, it deletes the old pod and starts a new pod
```sh
➜  ~ k -n kube-system get po
NAME                                              READY   STATUS    RESTARTS   AGE
coredns-558bd4d5db-c4sxf                          0/1     Pending   0          15m
coredns-558bd4d5db-x7wlq                          0/1     Pending   0          15m
etcd-kwok-kwok-control-plane                      1/1     Running   0          15m
kindnet-97jkb                                     1/1     Running   0          13m
kindnet-zrtg6                                     1/1     Running   0          15m
kube-apiserver-kwok-kwok-control-plane            1/1     Running   0          15m
kube-controller-manager-kwok-kwok-control-plane   1/1     Running   0          15m
kube-proxy-2f2mg                                  1/1     Running   0          15m
kube-proxy-hclf7                                  1/1     Running   0          13m
kube-scheduler-kwok-kwok-control-plane            1/1     Running   0          15m
kwok-controller-kwok-kwok-control-plane           1/1     Running   0          15m
➜  ~ docker exec -it kwok-kwok-control-plane mv /etc/kubernetes/manifests/kube-controller-manager.yaml /etc/kubernetes/manifests/kube-controller-manager.yaml.bak
➜  ~ k -n kube-system get po
NAME                                              READY   STATUS    RESTARTS   AGE
coredns-558bd4d5db-c4sxf                          0/1     Pending   0          17m
coredns-558bd4d5db-x7wlq                          0/1     Pending   0          17m
etcd-kwok-kwok-control-plane                      1/1     Running   0          17m
kindnet-97jkb                                     1/1     Running   0          16m
kindnet-zrtg6                                     1/1     Running   0          17m
kube-apiserver-kwok-kwok-control-plane            1/1     Running   0          17m
kube-controller-manager-kwok-kwok-control-plane   1/1     Running   0          29s
kube-proxy-2f2mg                                  1/1     Running   0          17m
kube-proxy-hclf7                                  1/1     Running   0          16m
kube-scheduler-kwok-kwok-control-plane            1/1     Running   0          17m
kwok-controller-kwok-kwok-control-plane           1/1     Running   0          17m
```

**Review by wzshiming — APPROVED** [review]

**Pull request merged**